community/primitive-collections/src/main/java/org/neo4j/concurrent/WorkSync.java

```diff
@@ -158,29 +191,127 @@ private W combine( WorkUnit<Material,W> batch )
             {
                 result = result.combine( batch.work );
             }
-            batch = batch.next;
+
+            do
+            {
+                batch = batch.next;
+            }
+            while ( batch.isCancelled() );
         }
         return result;
     }
 
     private void markAsDone( WorkUnit<Material,W> batch )
     {
+        // We will mark the first work unit as the successor - the thread in charge of unparking
+        // the other blocked threads - before we begin marking the units as done.
+        // This way, if the successor wakes up before we finish marking the units as done, it will
+        // know that it should wait and complete its task. In fact, it may trail our done-marking
+        // process and unpark threads in parallel.
+        batch.setAsSuccessor();
+        WorkUnit<Material,W> successor = batch;
         while ( batch != stackEnd )
         {
-            batch.done = true;
+            batch.complete();
             batch = batch.next;
         }
     }
 
-    private static class WorkUnit<Material, W extends Work<Material,W>>
+    private static class WorkUnit<Material, W extends Work<Material,W>> extends AtomicInteger
     {
+        static final int STATE_QUEUED = 0;
+        static final int STATE_PARKED = 1;
+        static final int STATE_DONE = 2;
+        static final int STATE_CANCELLED = 3;
+
         final W work;
+        final WorkUnit<Material,W> stackEnd;
+        final Thread owner;
         volatile WorkUnit<Material,W> next;
-        volatile boolean done;
+        volatile boolean successor;
 
-        private WorkUnit( W work )
+        private WorkUnit( W work, Thread owner, WorkUnit<Material,W> stackEnd )
         {
             this.work = work;
+            this.owner = owner;
+            this.stackEnd = stackEnd;
+        }
+
+        void park( long time, TimeUnit unit )
+        {
+            if ( compareAndSet( STATE_QUEUED, STATE_PARKED ) )
+            {
+                LockSupport.parkNanos( unit.toNanos( time ) );
+                compareAndSet( STATE_PARKED, STATE_QUEUED );
+            }
+            if ( successor )
+            {
+                // It's our job to go through all the 'next' references, and unpark all the threads
+                // for the work units that are marked as done.
+                // If we encounter a unit that isn't marked as done, then we have raced with the
+                // done-marking, and we just spin until we observe the done flag. This is pretty
+                // unlikely, since unparking a thread is pretty expensive on most platforms.
+                // We also spin on null 'next' references, in the unlikely event that our thread
+                // yet observed the complete enqueueing of work units prior to ours.
+                wakeUpAsSuccessor();
+            }
+        }
+
+        private void wakeUpAsSuccessor()
```

Comment:
I'd like it if this method had the same type of comments as other places about why it spins waiting for things.